### PR TITLE
[RzIL] Add Validator (Type-checker)

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Well-typed programs cannot go wrong.
  -- Robin Milner

This adds the IL validator, which performs static type-checking of both pure and effect ops among other checks. In particular, assuming the validator is correctly implemented, if it considers an op to be valid under some context, evaluating the op in the vm will never yield a runtime error, that is, an error where the vm itself errors, not an expected error state of the code being executed.
In our case, this includes for example:
* Any kind of type error: Conditions not being bool, bitvector sizes not matching, ...
* Variables not being available when they are accessed
* Using local variables with multiple different types in a single effect
* etc.

Any code that we lift must obey these rules. Thus, any analysis can rely on it.
The plan for this is to use the validator primarily in testing, development of lifters and for IL code coming from the outside.
If our lifting code is covered well enough by tests using the validator, we can omit the validation at runtime.

The only ops that do not have well-defined validation yet are `blk` and `goto` since their semantics, in particular regarding label handling are still a bit vague.
This also removes the `concat` and `unk` ops since they are unimplemented and not needed.

**Test plan**

See the unit tests.

### CODE REVIEW

File: librz/il/il_export.c [hunk 1]
```diff
@@ -844,3 +824,89 @@ RZ_API void rz_il_event_json(RZ_NONNULL RzILEvent *evt, RZ_NONNULL PJ *pj) {
 	free(tmp1);
 	free(tmp2);
 }
+
+RZ_API RZ_NONNULL const char *rz_il_op_pure_code_stringify(RzILOpPureCode code) {
```
Comment (XVilka) [inline librz/il/il_export.c:832]: Add a brief Doxygen please

File: librz/il/il_export.c [hunk 2]
```diff
@@ -844,3 +824,89 @@ RZ_API void rz_il_event_json(RZ_NONNULL RzILEvent *evt, RZ_NONNULL PJ *pj) {
...
+		return "load";
+	case RZ_IL_OP_LOADW:
+		return "loadw";
+	case RZ_IL_OP_PURE_MAX:
+		break;
+	}
+	return "invalid";
+}
+
+RZ_API char *rz_il_sort_pure_stringify(RzILSortPure sort) {
```
Comment (XVilka) [inline librz/il/il_export.c]: Same

File: librz/il/il_validate.c [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+
+static void var_kv_unown_free(HtPPKv *kv) {
+	free(kv->key);
+}
+
+/**
+ * Create a new global context for validation
+ * Vars and mems can be added manually with rz_il_validate_global_context_add_* functions.
+ */
+RZ_API RzILValidateGlobalContext *rz_il_validate_global_context_new_empty(ut32 pc_len) {
```
Comment (XVilka) [inline librz/il/il_validate.c:32]: `pc_len` - could it be zero? Are there any invalid values? Maybe add checks for them?

File: librz/il/il_validate.c [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+		free(ctx);
+		return NULL;
+	}
+	return ctx;
+}
+
+/**
+ * Define a new global variable in \p ctx
+ */
+RZ_API void rz_il_validate_global_context_add_var(RzILValidateGlobalContext *ctx, const char *name, RzILSortPure sort) {
```
Comment (XVilka) [inline librz/il/il_validate.c]: Please add `RZ_NONNULL` to the `name`

File: librz/il/il_validate.c [hunk 5]
```diff
@@ -0,0 +1,798 @@
...
+	rz_return_if_fail(ctx && name);
+	RzILSortPure *hts = RZ_NEW(RzILSortPure);
```
Comment (XVilka) [inline librz/il/il_validate.c:58]: Check for `hts == NULL`.

File: librz/il/il_validate.c [hunk 6]
```diff
@@ -0,0 +1,798 @@
...
+	}
+	return true;
+}
+
+static void local_context_fini(LocalContext *ctx) {
+	ht_pp_free(ctx->local_vars_known);
+	ht_pp_free(ctx->local_vars_available);
+}
+
+static bool local_var_copy_known_cb(void *user, const void *k, const void *v) {
```
Comment (XVilka) [inline librz/il/il_validate.c]: `user` is being dereferenced later, please add `RZ_NONNULL`

File: librz/il/il_validate.c [hunk 7]
```diff
@@ -0,0 +1,798 @@
...
+	RzILSortPure *sort = RZ_NEW(RzILSortPure);
+	if (!sort) {
+		return false;
+	}
+	*sort = *(RzILSortPure *)v;
+	ht_pp_update(dst->local_vars_known, k, sort);
+	return true;
+}
+
+static bool local_var_copy_avail_cb(void *user, const void *k, const void *v) {
```
Comment (XVilka) [inline librz/il/il_validate.c]: Same.

File: librz/il/il_validate.c [hunk 8]
```diff
@@ -0,0 +1,798 @@
...
+typedef struct {
+	bool failed;
+	RzStrBuf *report_builder;
+	const char *op_name;
+	LocalContext *dst;
+	LocalContext *src;
+} LocalContextMeet;
+
+// called on src, take the union of the known types or fail if they don't agree
+static bool local_var_meet_known_cb(void *user, const void *k, const void *v) {
```
Comment (XVilka) [inline librz/il/il_validate.c]: Same

File: librz/il/il_validate.c [hunk 9]
```diff
@@ -0,0 +1,798 @@
...
+		}
+		*dst_sort = src_sort;
+		ht_pp_update(meet->dst->local_vars_known, k, dst_sort);
+	}
+	return true;
+}
+
+// called on dst, remove all vars from dst that do not appear in src (intersection)
+static bool local_var_meet_avail_cb(void *user, const void *k, const void *v) {
+	LocalContextMeet *meet = user;
```
Comment (XVilka) [inline librz/il/il_validate.c:207]: Same

File: librz/il/il_validate.c [hunk 10]
```diff
@@ -0,0 +1,798 @@
...
+	};
+	ht_pp_foreach(b->local_vars_known, local_var_meet_known_cb, &meet);
+	if (meet.failed) {
+		return false;
+	}
+	ht_pp_foreach(a->local_vars_available, local_var_meet_avail_cb, &meet);
+	return true;
+}
+
+/////////////////////////////////////////////////////////
```
Comment (XVilka) [inline librz/il/il_validate.c:244]: You could use grouping feature by Doxygen with `\@{`, etc

File: librz/il/il_validate.c [hunk 11]
```diff
@@ -0,0 +1,798 @@
...
+ * of the same name when searching from the head.
+ */
+typedef struct local_pure_var_t {
+	const char *name;
+	RzILSortPure sort;
+	struct local_pure_var_t *next;
+} LocalPureVar;
+
+#define VALIDATOR_PURE_ARGS \
+	RZ_NULLABLE RzILOpPure *op, \
```
Comment (XVilka) [inline librz/il/il_validate.c:264]: Broken indentation?

File: librz/il/il_validate.c [hunk 12]
```diff
@@ -0,0 +1,798 @@
...
+		RZ_NONNULL RzILSortPure *sort_out, \
+		RZ_NONNULL RzStrBuf *report_builder, \
+		RZ_NONNULL const LocalContext *ctx, \
+		RZ_NULLABLE LocalPureVar *local_pure_var_stack
+static bool validate_pure(VALIDATOR_PURE_ARGS);
+typedef bool (*ValidatePureFn)(VALIDATOR_PURE_ARGS);
+#define VALIDATOR_PURE_NAME(op) validate_pure_##op
```
Comment (XVilka) [inline librz/il/il_validate.c:274]: Maybe a newline here? To make it more readable

File: librz/util/str_trim.c [hunk 13]
```diff
@@ -109,10 +109,12 @@ RZ_API void rz_str_trim_head(char *str) {
...
 
-// Remove whitespace chars from the tail of the string, replacing them with
-// null bytes. The string is changed in-place.
-RZ_API void rz_str_trim_tail(char *str) {
-	rz_return_if_fail(str);
+/**
+ * Remove whitespace chars from the tail of the string, replacing them with null bytes. The string is changed in-place.
+ * \return the string itself
+ */
+RZ_API char *rz_str_trim_tail(char *str) {
```
Comment (XVilka) [inline librz/util/str_trim.c]: Add then `RZ_NONNULL` here too.

Comment (XVilka) [reply librz/util/str_trim.c]: And ownership mark of the return, `RZ_BORROW`

File: librz/il/il_validate.c [hunk 14]
```diff
@@ -0,0 +1,798 @@
...
+		*sort_out = sort;
+	}
+	if (report_out) {
+		*report_out = rz_strbuf_is_empty(&report_builder) ? NULL : rz_str_trim_tail(rz_strbuf_drain_nofree(&report_builder));
+	}
+	rz_strbuf_fini(&report_builder);
+	return valid;
+}
+
+/////////////////////////////////////////////////////////
```
Comment (XVilka) [inline librz/il/il_validate.c:610]: Same, could use Doxygen grouping.

File: librz/il/il_validate.c [hunk 15]
```diff
@@ -0,0 +1,798 @@
...
+VALIDATOR_PURE(inv) {
+	RzILOpArgsBoolInv *args = &op->op.boolinv;
+	RzILSortPure sort;
+	VALIDATOR_DESCEND(args->x, &sort);
+	VALIDATOR_ASSERT(sort.type == RZ_IL_TYPE_PURE_BOOL, "Operand of boolean inv op is not boolean.\n");
+	*sort_out = rz_il_sort_pure_bool();
+	return true;
+}
+
+VALIDATOR_PURE(bool_binop) {
```
Comment (XVilka) [inline librz/il/il_validate.c:406]: For me, it's obvious that we are dealing with binary ops, then unary ops, etc. But for new devs it's probably worth to add a brief description for those validation functions.

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks very good. LGTM once you address my nitpicks.

File: librz/il/il_validate.c (hunk 3 above)
Comment (thestr4ng3r) [reply librz/il/il_validate.c:32]: Should indeed not be zero.

File: librz/il/il_validate.c (hunk 11 above)
Comment (thestr4ng3r) [reply librz/il/il_validate.c:264]: clang-format's fault

**Comment by thestr4ng3r** [comment]
I don't see what is supposed to be red there.

**Review by XVilka — APPROVED** [review]

**Comment by wargio** [comment]
i wanted to check this before merging

**Pull request merged**